API routes (and SSR pages?) with dynamic bundle chunks are not getting synced or bundled into a single file

### What happened?

Observed in [this repo](https://github.com/AnalogStudiosRI/api/pull/44/files#diff-015a06c6d91af296b9c65e45c3b0b16e362597112a198e0b4f6a59d7de9b7248) (that is using AWS SDK as an example) that when building and bundling the API routes, _src/pages/api/webhook/contentful.ts_ was emitting a LOT of chunk files
<img width="1541" alt="Image" src="https://github.com/user-attachments/assets/1644c780-d7fc-4a80-882f-fc41ee341fa7" />


This resulted in missing asset errors when uploaded to AWS Lambda
```sh
2025-03-14T16:46:38.181Z undefined ERROR Uncaught Exception {"errorType":"Error","errorMessage":"Cannot find module '/var/task/webhook-contentful.CCGsskaO.js' imported from /var/task/webhook-contentful.js","code":"ERR_MODULE_NOT_FOUND","url":"file:///var/task/webhook-contentful.CCGsskaO.js","stack":["Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/webhook-contentful.CCGsskaO.js' imported from /var/task/webhook-contentful.js"," at new NodeError (node:internal/errors:405:5)"," at finalizeResolution (node:internal/modules/esm/resolve:332:11)"," at moduleResolve (node:internal/modules/esm/resolve:1029:10)"," at moduleResolveWithNodePath (node:internal/modules/esm/resolve:876:12)"," at defaultResolve (node:internal/modules/esm/resolve:1255:79)"," at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:404:12)"," at ModuleLoader.resolve (node:internal/modules/esm/loader:373:25)"," at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:250:38)"," at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:39)"," at link (node:internal/modules/esm/module_job:75:36)"]}
```

because these additional chunk assets were not being managed / known to the adapter plugin and thus not uploaded
<img width="1414" alt="Image" src="https://github.com/user-attachments/assets/499d86f4-3f06-46a2-aab5-4d0e6cbfcbf9" />

----

This seems to contradict the expectation of https://github.com/ProjectEvergreen/greenwood/issues/1118, in which in theory, there should have only created on output file anyway?

### Steps to reproduce

1. Create an API route
1. Install / add the AWS SDK to a route
1. Run the build

(haven't tested for SSR pages, but presumably the same thing could / would happen since they share the same Rollup configuration options)

### Environment

- Latest (or Alpha) version of Greenwood
- Have an (any) adapter plugin

### Additional Context

Doing a little research, looks like per [this issue](https://github.com/rollup/rollup/issues/2616), I had the Rollup configuration wrong.  Looks like I should have been only using
- `output.file` (no `dir`)
- `inlineDynamicImports`

Will also be curious if this somehow fixes / better handles [`new URL`](https://github.com/ProjectEvergreen/greenwood/discussions/1204) usage?  And maybe we don't need to track assets anymore?

--- a/packages/cli/package.json
+++ b/packages/cli/package.json
@@ -68,6 +68,7 @@
     "wc-compiler": "~0.16.0"
   },
   "devDependencies": {
+    "@aws-sdk/client-cloudfront": "^3.726.0",
     "@mapbox/rehype-prism": "^0.5.0",
     "@shoelace-style/shoelace": "^2.18.0",
     "@spectrum-css/card": "^9.3.0",

--- a/packages/cli/src/config/rollup.config.js
+++ b/packages/cli/src/config/rollup.config.js
@@ -180,6 +180,21 @@ function greenwoodSyncSsrEntryPointsOutputPaths(compilation) {
   };
 }
 
+// https://github.com/ProjectEvergreen/greenwood/issues/1465
+// https://github.com/ProjectEvergreen/greenwood/discussions/1204#discussioncomment-12582424
+function getAllBundleChunks(key, bundles) {
+  const chunks = [];
+
+  for (const bundle in bundles) {
+    const { fileName } = bundles[bundle];
+    if (fileName !== key && fileName.startsWith(key.replace(".js", ""))) {
+      chunks.push(bundles[bundle].fileName);
+    }
+  }
+
+  return chunks;
+}
+
 function greenwoodSyncApiRoutesOutputPath(compilation) {
   return {
     name: "greenwood-sync-api-routes-output-paths",
@@ -198,10 +213,14 @@ function greenwoodSyncApiRoutesOutputPath(compilation) {
 
           if (compilation.manifest.apis.has(route)) {
             const api = compilation.manifest.apis.get(route);
+            const linkedAssets = getAllBundleChunks(key, bundle).map(
+              (chunk) => new URL(`./api/${chunk}`, outputDir).href,
+            );
 
             compilation.manifest.apis.set(route, {
               ...api,
               outputHref: new URL(`./api/${key}`, outputDir).href,
+              assets: api.assets ? [...api.assets, ...linkedAssets] : linkedAssets,
             });
           }
         }
@@ -716,7 +735,6 @@ const getRollupConfigForBrowserScripts = async (compilation) => {
 };
 
 const getRollupConfigForApiRoutes = async (compilation) => {
-  // console.log('getRollupConfigForApiRoutes', compilation.manifest.apis.values())
   const { outputDir } = compilation.context;
 
   return [...compilation.manifest.apis.values()]

--- a/packages/cli/test/cases/build.plugins.adapter/build.config.plugins-adapter.spec.js
+++ b/packages/cli/test/cases/build.plugins.adapter/build.config.plugins-adapter.spec.js
@@ -28,12 +28,15 @@
  *       nested/
  *         endpoint.js
  *       greeting.js
+ *       webhook.js
  *     blog/
  *       first-post.js
  *       index.js
+ *     about.js
  *     index.js
  */
 import chai from "chai";
+import glob from "glob-promise";
 import path from "path";
 import { getOutputTeardownFiles } from "../../../../../test/utils.js";
 import { Runner } from "gallinago";
@@ -199,6 +202,32 @@ describe("Build Greenwood With: ", function () {
 
         expect(heading).to.have.lengthOf(1);
         expect(heading[0].textContent).to.equal("Nested SSR First Post page should work!");
+      });
+    });
+
+    // https://github.com/ProjectEvergreen/greenwood/issues/1465
+    describe("Adapting of an API route with lots of dynamic chunk bundles", function () {
+      let chunks;
+
+      before(async function () {
+        chunks = await glob.promise(path.join(outputPath, "adapter-output/webhook.*.js"));
+      });
+
+      it("should contain one javascript file in the output directory", async function () {
+        expect(chunks.length).to.equal(15);
+      });
+    });
+
+    // https://github.com/ProjectEvergreen/greenwood/issues/1465
+    describe("Adapting of a SSR page function with lots of dynamic chunk bundles", function () {
+      let chunks;
+
+      before(async function () {
+        chunks = await glob.promise(path.join(outputPath, "adapter-output/about.route.chunk.*.js"));
+      });
+
+      it("should contain one javascript file in the output directory", async function () {
+        expect(chunks.length).to.equal(16);
       });
     });
   });

--- a/packages/cli/test/cases/build.plugins.adapter/generic-adapter.js
+++ b/packages/cli/test/cases/build.plugins.adapter/generic-adapter.js
@@ -1,4 +1,5 @@
 import fs from "fs/promises";
+import path from "path";
 import { checkResourceExists } from "../../../../cli/src/lib/resource-utils.js";
 
 function generateOutputFormat(id, type) {
@@ -20,6 +21,7 @@ function generateOutputFormat(id, type) {
 }
 
 async function genericAdapter(compilation) {
+  const { outputDir } = compilation.context;
   const adapterOutputUrl = new URL("./adapter-output/", compilation.context.projectDirectory);
   const ssrPages = compilation.graph.filter((page) => page.isSSR);
   const apiRoutes = compilation.manifest.apis;
@@ -31,15 +33,32 @@ async function genericAdapter(compilation) {
   for (const page of ssrPages) {
     const { id } = page;
     const outputFormat = generateOutputFormat(id, "page");
+    const chunks = (await fs.readdir(outputDir)).filter(
+      (file) => file.startsWith(`${id}.route.chunk`) && file.endsWith(".js"),
+    );
 
     await fs.writeFile(new URL(`./${id}.js`, adapterOutputUrl), outputFormat);
+
+    for (const chunk of chunks) {
+      await fs.cp(new URL(`./${chunk}`, outputDir), new URL(`./${chunk}`, adapterOutputUrl), {
+        recursive: true,
+      });
+    }
   }
 
   for (const [key] of apiRoutes) {
-    const { id } = apiRoutes.get(key);
+    const { id, assets } = apiRoutes.get(key);
     const outputFormat = generateOutputFormat(id, "api");
 
     await fs.writeFile(new URL(`./api-${id}.js`, adapterOutputUrl), outputFormat);
+
+    for (const asset of assets) {
+      const name = path.basename(asset);
+
+      await fs.cp(new URL(`./api/${name}`, outputDir), new URL(`./${name}`, adapterOutputUrl), {
+        recursive: true,
+      });
+    }
   }
 }
 

--- a/packages/cli/test/cases/build.plugins.adapter/src/pages/about.js
+++ b/packages/cli/test/cases/build.plugins.adapter/src/pages/about.js
@@ -1,0 +1,16 @@
+import * as AWS from "@aws-sdk/client-cloudfront";
+
+const CONFIG = {
+  region: "us-east-1",
+  accessKeyId: "process.env.AWS_ACCESS_KEY_ID",
+  secretAccessKey: "process.env.AWS_SECRET_ACCESS_KEY",
+  distributionId: "process.env.AWS_CLOUDFRONT_ID",
+};
+
+export default class AboutPage extends HTMLElement {
+  async connectedCallback() {
+    const cfClient = new AWS.CloudFront(CONFIG);
+
+    this.innerHTML = `<h1>About Page (${cfClient.initConfig.serviceId})</h1>`;
+  }
+}

--- a/packages/cli/test/cases/build.plugins.adapter/src/pages/api/webhook.js
+++ b/packages/cli/test/cases/build.plugins.adapter/src/pages/api/webhook.js
@@ -1,0 +1,31 @@
+import * as AWS from "@aws-sdk/client-cloudfront";
+
+const CONFIG = {
+  region: "us-east-1",
+  accessKeyId: "process.env.AWS_ACCESS_KEY_ID",
+  secretAccessKey: "process.env.AWS_SECRET_ACCESS_KEY",
+  distributionId: "process.env.AWS_CLOUDFRONT_ID",
+};
+
+export async function handler(request) {
+  const cfClient = new AWS.CloudFront(CONFIG);
+  const body = await request.json();
+  const entity = body?.sys?.contentType?.sys?.id || "";
+
+  await cfClient.createInvalidation({
+    DistributionId: CONFIG.distributionId ?? "",
+    InvalidationBatch: {
+      CallerReference: new Date().getTime().toString(),
+      Paths: {
+        Quantity: 1,
+        Items: [`/api/v2/${entity}s*`],
+      },
+    },
+  });
+
+  return new Response(JSON.stringify({ msg: "success" }), {
+    headers: new Headers({
+      "Content-Type": "application/json",
+    }),
+  });
+}

